### DISCUSSION TITLE
Pagination for Patient $everything bundle

### DISCUSSION THREAD

**Discussion by dylman123** [medplum/medplum#4000] [General]
Hey Medplum team, I am wondering what paths may be possible to receive a paginated bundle response for the Patient $everything operation? Seeking your guidance here.

## Why
* Today the Medplum implementation of $everything returns all data in the Patient compartment
* If the patient has many resources, this can be a large JSON response
* Pagination is supported for other searches by Medplum

## What
* One possibility is to return a subset of the $everything bundle based on `_count` and `_offset` params (possibly at [this line](https://github.com/medplum/medplum/blob/main/packages/server/src/fhir/operations/patienteverything.ts#L123)).
* However this would, in effect, be paginating in memory. And each subsequent page request to the $everything API route would require the server to repeat work (i.e. re-do the entire set of resource searches required to build the bundle).
* Can you think of a more efficient path than what I propose?
* I am happy to open a PR to suggest the feature, based on the discussion here.

**Comment by healthbjk** [comment]
Related issue: https://github.com/medplum/medplum/issues/3944

**Comment by codyebberson** [comment]
Hi @dylman123 - thanks for submitting.

> Can you think of a more efficient path than what I propose?

A more efficient solution is possible using `UNION` queries.

If you're familiar with SQL, imagine something like this:

```sql
SELECT * FROM (
  SELECT column1, column2, ... FROM "Patient"
  UNION
  SELECT column1, column2, ... FROM "Observation"
  UNION
  SELECT column1, column2, ... FROM "ServiceRequest"
) AS combined_results
LIMIT your_limit OFFSET your_offset;
```

Our plan is to implement that as a lower level primitive in base search with the `_type` parameter: https://github.com/medplum/medplum/issues/2021

Once that is available, then in-Postgres pagination should be trivial.

Adding support for the `_type` parameter is currently unscheduled work.  I can bump this with the team during our next planning session.

**Reply by dylman123** [reply]
Thanks @codyebberson. That makes sense

**Reply by codyebberson** [reply]
This was fixed in https://github.com/medplum/medplum/pull/4296

Released in Version 3.1.1 https://github.com/medplum/medplum/pull/4319
